### PR TITLE
Fix unreliable entities pushed by pistons

### LINKED ISSUE #5189: Armor stands in ice highway machines act different than in Vanilla
I was advised on Discord to open a GitHub issue for this since I'm not the first one with this problem, so here we are.

### What behaviour is expected:
On our small 1.16.5 Paper server we've tried to build the following redstone machine, designed by the SciCraft community:
[https://www.youtube.com/watch?v=6Hf62xgulzc](https://www.youtube.com/watch?v=6Hf62xgulzc)
This is a flying machine, which pushes armor stands with frost walker boots on them and simply builds a 4 block wide "highway" with railings on its side as it travels.

The problem arises in the behaviour of the armor stands in the machine. The base concept is, that a slime block is pushed from above into the upper half of each of the armor stands with a piston, and because of this, the flying machine drags them with itself when it moves. It's a somewhat popular design, and it works flawlessly in single player or on Vanilla servers (it was tested multiple times by us too when we kept failing to get it to work on Paper).

### What behaviour is observed:
When the machine is turned on on our Paper server, the armor stands "clip out" from the slime blocks, without moving, completely breaking the machine.

### Steps/models to reproduce:
The build schematic we used can be found in the description of this video:
[https://www.youtube.com/watch?v=5vRmQNTMdco](https://www.youtube.com/watch?v=5vRmQNTMdco)

It is also a block-by-block tutorial, showing the machine in work too.

### Plugin list:
- BetterSleeping 3.0.4
- CoreProtect 19.4
- DynMap 3.1-beta6
- EntityTrackerFixer 1.3.3
- EssentialsX, EssentialsX-Chat, EssentialsX-Spawn 2.18.2.0
- EternalLight 1.7
- LuckPerms 5.2.76
- PlaceholderAPI 2.10.9
- ProtocoLib 4.5.1
- SAML 2.0.0
- ServerListPlus 3.4.8
- SkinsRestorer 13.8.8
- TabList 5.5.0
- Vault 1.7.3
- WorldEdit 7.2.0

### Paper version:
`git-Paper-"1c8b6065e" (MC: 1.16.5)`

### Anything else:
I'm not really sure, but I'm glad to give any more information needed to find any solution for this issue.

### PR DESCRIPTION
**Pull request by petersv5**
[resubmission of a corrected pull as I accidentally closed the previous #6301]

This patch solves #5189 where entities in slime blocks are unreliable
in PaperMC. This is known to break some contraptions like the ice road
maker mentionend there.

The removal of 1/4 of all ticks for an entity that takes place in
ActivationRange.checkIfActive() results in an entity like e.g.
an armor stand not noticing that it has had setDeltaMovement() called.
The result is that an armor stand inside a slime block gets left behind
1/4 of the pushes.

There were already some hard coded logic in checkIfActive to not apply
the culling in some circumstances. The attach patch takes a slightly
different approach and adds a general mechanism where the entity knows
how many ticks into the future that it has to be fully ticked and
ActivationRange queries that.

Another approach would be to just have a "wasMovedByPiston" memory in
the entity and have the timing logic in ActivationRange. I prepared a
patch for that as well in case that is preferable. All things being
equal I thought it better to submit the more general patch if more cases
of needing to temporarily override the tick culling.

As dicussed on the discord channel this patch does not affect the
configration option to not tick armor stands as that is handled
in ArmorStand.tick().

### CODE REVIEW

**Review by qixils — APPROVED** [review]

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: What's with the extra spaces all over the place?

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: I'd prefer we make this private with a getter; `requestExactTicks` already acts as a setter, so might as well do proper encapsulation.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Have you tested with lower values?

**Review by Proximyst — CHANGES_REQUESTED** [review]
Blocking on answer for lower values; the rest could be handled pretty easily by the merger.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Which extra spaces are you referring to? The ones inside the //Paper comments or are there others I should fix? 
Should I fix them on my side?

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Ok, not a problem. I started by adding getter/setters, then pulled those out following a pattern I saw in other places. Easy enough to put back.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: I belive that the moveCollidedEntities is also called for "bounced", and I thought it best to give enough time for the movement to die down. For the stuck entities 2 ticks should be enough. I'll test that, as well if bounced entities are affected by this patch. Back in 10 minutes.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Yeah, the ones inside the paper comments. If you feel like fixing them, sure, otherwise we'll take care of it on merge (though it'd be kindest if you did 😄).

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: If you have some source other places where you pulled the number, feel free to keep it, but add a comment on the line that has said number.

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: After a bit of testing: for items stuck inside a slime block 2 ticks is sufficient to guarantee it starts and stops moving. 
This function is also called for non-stuck items, that are laucnhed ballistically. I think they will move differently depending on if they get all ticks or not while in motion. Only 2 ticks are needed for the ice maker and similar machines based on stuck entities.

On the other hand, I think paper already accepts non-vanilla behaviour for dynamic movement as the 1/4 tick removal exists, so maybe we should lower tihs value to 2? We can always make it configurable in the future if good test cases are found. I suspect that perhaps tnt ender pearl cannons would be impacted by this, but I have no good test case for this.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Updated in the new commit

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Done in the new commit. Let me know if you perfer a different naming convention.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Updated with a chnage to 2 ticks only. This was verfied as sufficient both manually and with the Ilmango deisgn from issue 5189 running over 1200 blocks without issue. Bounced item trajectories are leaft for the future, once someone designs a test case for comparing differences between vanilla and paper.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Might as well use `this.isTickRequired` here.

Comment (Proximyst) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Also prefer using `this.` prefixes; not at all a blocker, however.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: You have several instances of this :)

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: I suppose a test for some ender pearl device could be done to make sure they remain functional? Otherwise, I'm sure the various broken machines will turn up in the issue tracker after merge, anyway.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Done in new commit.

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Done in new commit.

**Comment by petersv5** [comment]
I think someone needs to approve the integration test workflow (presumably to verify that the patch applies?)

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: I fixed the two instances of extra spaces I found. Were there any more?

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Any machines working before the fix should see no change after the fix, excpet perhaps a tiny bit less variation. It is likely still not vanilla behaviour for bounced entities, but I would need a test case before we delve into that. I suggest we leave that for a future issue...

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Proximyst) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: This and the below hunk need reformatting

**Review by Proximyst — APPROVED** [review]

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Will you handle that as a part of the merge, or do you want me to fix it and resubmit?

Comment (Proximyst) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Either or; your approval only gets removed if a larger change comes

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: I'll be away for the rest of the day.  If it is not already merged when I get back I can update the formatting.

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: I am not sure what formatting is requested. Should I switch to the multi-line format for the // Paper marker, or was it something else you had in mind?

Comment (qixils) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: i think she means the 4 spaces before the comment. everything else in paper just uses 1 (`; // Paper - `)

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (NoahvdAa) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]:
```suggestion
+    // Paper end
```

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (NoahvdAa) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]:
```suggestion
+        // Paper start - immunize entities requesting exact ticks, e.g. being affected by a piston
```

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (qixils) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: since there's already a paper block above this, you can just move this into the other paper block (i.e. remove lines 32-34)

Comment (lexikiq) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:44]: also, since there's already a paper block above this, you can just move this into the other paper block (i.e. remove lines 32-34)

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (qixils) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: unnecessary, this is part of an existing paper block

Comment (NoahvdAa) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Whoops, didn't see that. Is there a way for me to mark this as solved, or is that something only @petersv5 can do?

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:74]: Oops, I approved that change. Let me commit a revert to that.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Updated in new commit

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Updated. let me know if I got this right.

**Comment by petersv5** [comment]
I think the last commit takes care of all review comments so far. 
It seems that a maintainer has to approve the integration test workflow again though.

**Review by NoahvdAa — APPROVED** [review]
LGTM!

**Review by qixils — APPROVED** [review]

**Comment by petersv5** [comment]
Is there anything else I should do before this is ready to go?

**Review by electronicboy — APPROVED** [review]

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (Machine-Maker) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Personally, for counters I like using `--` instead of ` -= `, I think it just signals this is counting smth easily.

**Review by Proximyst — APPROVED** [review]
@kennytv :dinkdonk:

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Want me to update, or will you do that during merge?

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Updated as requested.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (kennytv) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: 2 is not enough (and use the current tick time +). Also add `this.activatedTick = MinecraftServer.currentTick + <whatever would be appropriate>;`, probably 10-20

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (kennytv) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Make this public, move it under `activatedTick`, store a ticktime, remove getter/setter, and preferably rename this to something like `activationImmunityTick`.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (kennytv) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Move this into the move method where `this.activatedTick` is assigned (and use the current tick time + 20)

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (kennytv) [inline patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Directly check against the public immunity tick time ~~and move the check into `ActivationRange#checkEntityImmunities`~~

**Review by kennytv — CHANGES_REQUESTED** [review]
Put this into the EAR patch (0364) if you can please. Most of the requested changes are for a cleaner diff. And please make sure if properly works after that one more time, I only briefly tested the ice machine linked in the issue.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Do you want me to switch from tracking/counting down) a certain number of ticks to instead track the last tick to be immune in absolute terms?

The patch originally started with 10 ticks, but Proimyst asked it to be reduced. 2 ticks are sufficient for entities stuck in pistons, as they will complete the movment in 2 ticks from reading the code and testing. The original 10 ticks were to allow for entities being bounced (not stuck inside slime) to complete ballisitc movement. After some back and forth we decided to drop that as we had no test cases for how long was sufficient anyway.  Do you want me to up the immunity again to 20 ticks (one second)?

Whis is the semantic for the this.activatedTick?

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Oki. I was asked to make it private (was origianlly public) and introduce getters/setters. Are you sure you want this to be direct access instead of setter/getter?

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (kennytv) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: If you are 100% certain 2 ticks are enough, then that's fine. In my testing it wasn't but I had some other code changes flying around as well, so idk. activatedTick is basically exactly what you are doing without the force part, since it additionally does that skip every 4 ticks if not immune. so if activatedTick >= currentTick, the entity should be ticked (so you do the same as that, but checked in the immunity method).

I don't have the code open anymore, but if the immunity check method does not have any other usages, you can keep your check at the place it is currently at to return true as being active. Not sure if you can skip setting activatedTick tho, but it doesn't hurt eitherway

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (kennytv) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Yes, to make it more in-line with the original EAR patch once merged into it.

File: patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch
Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: I may not follow exactly here. Did you mean to both move activatedTime approriately into the future + also having an entity-centric immunization as in my patch but with updated semantics?

Comment (kennytv) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Yesn't to the first, since I wasn't sure. Looking at the activatedTick usages, you don't necessarily have to add the codeline I provided. Just keep the number at whatever you are sure works and only update the immunity `entity.activationImmunityTick = MinecraftServer.currentTick + <ticks it should be active for>;` (your countdown variable but made into absolute tick time)

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Yes, my understanding is that the entity is already activated in this particular case and that it is just the decimation-by-4 that is casuing the intermittent issues.

One thing strikes me: would it make more sense to say that activatedTick in the future should be immune to the decimation? I have not followed all the other uses of the activatedTick, maybe that would completely neuter the load improvement from the decimation?

Comment (kennytv) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: If activatedTick is behind the currentTick, the entity is marked as inactive and doesn't tick at all unless it is immune given by the immune check in ActivationRange. Sidenote: If *this* particular spot didn't exist and you only had to account for the other place in the move() method, you could even just use `entity.pistonDeltasGameTime` and shove it into the immunity method.. but you can't, since this is here

So the correct way of keeping it active is by "immunizing" it - activatedTick active entities are supposed to be skipped every 4th unless "immune". But since the immunity check method is used nowhere else, it doesn't matter wheter your check is put into that ActivationRange method or exactly where you have it at the moment (tho you *do* have to set activatedTick here *if* you moved it into the immune check method).

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: A little more digging on my part: Today the activation of entities moved by being stuck on honey or bounced _on_ slime blocks (and some other cases) get activated fo 20 ticks in Entity.move() due to the movementType being MoverType.PISTON. For items that are stuck _in_ slime instead get moved by Entity.setDeltaMovement that keeps being updated by the PistonMovingBlockEntity. 

In my patch I did the immunization for both cases in PistonMovingBlockEntity. However, maybe the immunzation for the non-stuck-in-slime should be moved next to the activatedTime update in Entity.move() ?

What concerns me is that I cannot find where the entities stuck inside slime received their activation before my patch. They were clearly activated as the failure rate was 1/4. Was it due to short distance to the player? I propose to add setting both activatedTick and actiavtedImmunityTick when doing the setDeltaMovememnt call and moving the test into the checkEntityimmunities() as you suggested.

Comment (kennytv) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Yep, that should work. 👍

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Is it ok to blindly assign activatedTick / activatedimmunityTick with currentTick+xxxx, or sohuld I use max(activatedtick, currentTick+xx) ?

Comment (kennytv) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: Good point. The piston type in the move method currently is the only one that would theoretically not recover itself, but better save then sorry

Comment (petersv5) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: MinecraftServer does not seem to be used in PistonMovingBlockEntity at the moment. What is the best way to get at the currentTick there? Should I just import it, or is there some other way?

Comment (kennytv) [reply patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch]: I believe it is a static value, so using a fully qualified import, `net.minecraft.server.MinecraftServer.currentTick`

**Comment by petersv5** [comment]
Revamped the patch as requested and merged into the EAR patch.
Immunization upped to 10 ticks for stuck-in-slime entities out of an abundance of caution.
Tested with the ice road maker for hundreds of blocks.

File: patches/server/0364-Entity-Activation-Range-2.0.patch
Comment (kennytv) [inline patches/server/0364-Entity-Activation-Range-2.0.patch:591]: What happened here 👀

**Comment by petersv5** [comment]
gradlew rebuildPatches touched several patches after the EAR patch with changed indices. I guess I need to commit all of those as well after doing fixup on the EAR commit in Paper-Server?
I only commited and pushed the update 0364 patch.

**Comment by kennytv** [comment]
Yes, you have to commit everything

**Comment by petersv5** [comment]
I feel I am swimming in the deep end here.... This compiles and works when I clone this PR.
Want me to do a rebase to the current state as well?

File: patches/server/0364-Entity-Activation-Range-2.0.patch
Comment (kennytv) [inline patches/server/0364-Entity-Activation-Range-2.0.patch]: Do a fully qualified import here and remove the class import

**Review by kennytv — CHANGES_REQUESTED** [review]
Don't worry about the failed build, that has nothing do with the pr. Rebasing would be nice - not sure if it already fixes the build, but if not, the merger will rebase again.

**Comment by petersv5** [comment]
Removed the import + rebased.

**Review by kennytv — APPROVED** [review]
Thank you! We'll see about merging once the build issues are fixed

**Comment by petersv5** [comment]
> Thank you! We'll see about merging once the build issues are fixed

Want me to do anything on my end? 
The build works for me when I clone the PR and do applyPatches + reobfJar.

**Comment by petersv5** [comment]
I think the empty commit reran the integration test successfully.

**Comment by petersv5** [comment]
What is the status of this PR? Anything left to be updated?

**Comment by Proximyst** [comment]
We’ll look at merging it whenever we have time and opportunity to do so. It takes time, sadly.

**Pull request merged**
